### PR TITLE
Add binding redirect generation for desktop tests

### PR DESCRIPTION
**Pull request by karajas**
/cc @danmosemsft @AlexGhiondea @ericstj @weshaggard

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            XDocument doc = new XDocument(new XElement("configuration", new XElement("runtime", bindingRedirectAssemblies)));
+            foreach (ITaskItem executable in Executables)
+            {
+                string executableName = Path.GetFileName(executable.ItemSpec);
+                using (FileStream fs = new FileStream(Path.Combine(OutputPath, executableName+".config"), FileMode.Create))
+                {
+                    doc.Save(fs);
+                }
+            }
+            return true;
```
Comment (danmosemsft) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: as general good practice Execute() on all MSBuild tasks should return thus
`return !Log.HasLoggedErrors;`

In this case it will be true anyway -- but are there any exceptions/error cases to handle and log here?

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        public ITaskItem[] AssemblyIdentities { get; set; }
+
+        [Required]
+        public ITaskItem[] Executables { get; set; }
+
+        [Required]
+        public string OutputPath { get; set; }
+
+        private static XNamespace ns { get; set; }
+        public override bool Execute()
```
Comment (danmosemsft) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: line spacing

**Comment by karajas** [comment]
Fixes issue https://github.com/dotnet/buildtools/issues/1092

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets [hunk 3]
```diff
@@ -206,6 +209,28 @@
       OutputListLocation="$(OutDir)assemblylist.txt"
       Condition="'$(UseLegacyXunitPerfRunner)'=='true'"
      />
+     
+     <!-- For Desktop Tests, we need to generate Binding Redirects. -->
+     <ItemGroup>
+        <_FileNamesInReferenceCopyLocalPaths Include="@(ReferenceCopyLocalPaths->'%(Filename)%(Extension)')">
+          <SourcePath>%(ReferenceCopyLocalPaths.Identity)</SourcePath>
+        </_FileNamesInReferenceCopyLocalPaths>
+        <_FileNamesInTestCopyLocalByFileNameWithoutDuplicates Include="@(_TestCopyLocalByFileNameWithoutDuplicates->'%(Filename)%(Extension)')">
```
Comment (danmosemsft) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets]: Isn't this 
```
+        <_FileNamesInTestCopyLocalByFileNameWithoutDuplicates Include="@(_TestCopyLocalByFileNameWithoutDuplicates->'%(Filename)%(Extension)')">
 +          <SourcePath>%(_TestCopyLocalByFileNameWithoutDuplicates.SourcePath)</SourcePath>
 +        </_FileNamesInTestCopyLocalByFileNameWithoutDuplicates>
```

the same as 
```
<_FileNamesInTestCopyLocalByFileNameWithoutDuplicates  Include="@(_TestCopyLocalByFileNameWithoutDuplicates)"/>
```

ie., metadata like `SourcePath` should be copied automatically when you populate one list from anotehr using an item transform (an @ expression)

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets [hunk 4]
```diff
@@ -206,6 +209,28 @@
...
+          <SourcePath>%(_TestCopyLocalByFileNameWithoutDuplicates.SourcePath)</SourcePath>
+        </_FileNamesInTestCopyLocalByFileNameWithoutDuplicates>
+        <_FilteredListForBindingRedirect Include="@(_FileNamesInReferenceCopyLocalPaths)" Exclude="@(_FileNamesInTestCopyLocalByFileNameWithoutDuplicates)" />
+        <_FilteredListForBindingRedirect Include="@(_FileNamesInReferenceCopyLocalPaths)" />
```
Comment (danmosemsft) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets]: I'm not sure what this accomplishes
```
+        <_FilteredListForBindingRedirect Include="@(_FileNamesInReferenceCopyLocalPaths)" Exclude="@(_FileNamesInTestCopyLocalByFileNameWithoutDuplicates)" />
 +        <_FilteredListForBindingRedirect Include="@(_FileNamesInReferenceCopyLocalPaths)" />
```

It includes one list excepting some things, then includes it again entirely.

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            ns = "urn:schemas-microsoft-com:asm.v1";
+            XElement bindingRedirectAssemblies = new XElement(ns + "assemblyBinding");
+            foreach (ITaskItem assembly in AssemblyIdentities)
+            {
+                string assemblyVersion = assembly.GetMetadata("Version");
+                XElement assemblyIdentity = new XElement("assemblyIdentity",
+                                                         new XAttribute("name", assembly.GetMetadata("Name")),
+                                                         new XAttribute("publicKeyToken", assembly.GetMetadata("PublicKeyToken")),
```
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: why not do the same for PublicKeyToken and provide a default if we don't have one?

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                                                         new XAttribute("culture", string.IsNullOrEmpty(assembly.GetMetadata("Culture")) ? "neutral" : assembly.GetMetadata("Culture")));
```
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: Not sure how expensive `GetMetadata("Culture")` is, but if it is expensive, you should cache the call and then check its value.

Comment (danmosemsft) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: It's a lookup in a small dictionary so no need. Some built-in metadata like %(FullPath) is expensive, but it's cached by MSBuild in that dictionary.

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs (hunk 1 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: Will do so, probably a better idea anyways. I think the only other exceptions we can expect are possible missing metadata and xml write exceptions. Should I try/catch the whole thing? Don't think I can log and continue.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets (hunk 3 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets]: So ReferenceCopyLocalPaths do not have SourcePath, I just wanted to standardize the metadata that I wanted in both groups.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets (hunk 4 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets]: So this is the case where an assembly has been placed in the test path prior to resolving references, so I want to remove files from my list of files from resolve and include the ones that have been copied instead with the appropriate metadata.

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                string executableName = Path.GetFileName(executable.ItemSpec);
+                using (FileStream fs = new FileStream(Path.Combine(OutputPath, executableName+".config"), FileMode.Create))
+                {
+                    doc.Save(fs);
+                }
+            }
+            return true;
+        }
+
+        private XElement RemoveNamespace(XElement node)
```
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: Why is this needed?

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs (hunk 5 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: Sure, is there a source from which I can grab the default PublicKeyToken?

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs (hunk 6 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: Will fix.

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                                                        new XAttribute("newVersion", assemblyVersion));
+                bindingRedirect = RemoveNamespace(bindingRedirect);
+                XElement dependentAssembly = new XElement("dependentAssembly",
+                                                          assemblyIdentity,
+                                                          bindingRedirect);
+                dependentAssembly = RemoveNamespace(dependentAssembly);
+                bindingRedirectAssemblies.Add(dependentAssembly);
+            }
+            XDocument doc = new XDocument(new XElement("configuration", new XElement("runtime", bindingRedirectAssemblies)));
+            foreach (ITaskItem executable in Executables)
```
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: I don't believe any of the corefx tests are built in this fashion. If we do want to include a local app.config, maybe I can add a override that can be passed through the new test csproj.

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs (hunk 7 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: I was having some difficulty understanding why empty namespaces were being added to the elements. Will try to find a cleaner way.

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs (hunk 5 above)
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: I don't think so. I was thinking more about skipping if the assembly does not have a publickeytoken.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets [hunk 9]
```diff
@@ -206,6 +209,28 @@
...
+        <_FilteredAssembliesForBindingRedirect Include="%(_FilteredListForBindingRedirect.SourcePath)" Condition="$([System.String]::Copy('%(_FilteredListForBindingRedirect.SourcePath)').EndsWith('.dll')) or $([System.String]::Copy('%(Remaining.SourcePath)').EndsWith('.exe'))" />
```
Comment (danmosemsft) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets]: This String.Copy trick is so horrible I opened a suggestion https://github.com/Microsoft/msbuild/issues/1465

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets [hunk 10]
```diff
@@ -206,6 +209,28 @@
...
+        <ExecutablesForBindingRedirect Include="%(IncludedFileForRunnerScript.SourcePath)" Condition="$([System.String]::Copy('%(IncludedFileForRunnerScript.SourcePath)').EndsWith('.exe'))"/>
```
Comment (danmosemsft) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets]: Suggestion, instead of all these item manipulation expressions, is it possible to just give @(ReferenceCopyLocalPaths) to the task ? It can filter out all but exe and dll, and can consult %(SourcePath) if needed.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets (hunk 9 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets]: I could probably just compare on Extension in that case.

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                    string assemblyVersion = assembly.GetMetadata("Version");
+                    string assemblyName = assembly.GetMetadata("Name");
+                    string culture = string.IsNullOrEmpty(assembly.GetMetadata("Culture"))
+                        ? "neutral"
+                        : assembly.GetMetadata("Culture");
+                    XElement assemblyIdentity = new XElement("assemblyIdentity",
+                        new XAttribute("name", assemblyName),
+                        new XAttribute("publicKeyToken", publicKeyToken),
+                        new XAttribute("culture", culture));
+                    assemblyIdentity = RemoveNamespace(assemblyIdentity);
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: Use `ns +` to avoid this.

**Comment by karajas** [comment]
Removed deduping from task and fixed up publishtests.targets to have a variable with the right test assets. PTAL.

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+        public override bool Execute()
+        {
+            ns = "urn:schemas-microsoft-com:asm.v1";
+            XElement bindingRedirectAssemblies = new XElement(ns + "assemblyBinding");
+            List<XElement> dependentAssemblyList = new List<XElement>();
+            foreach (ITaskItem assembly in AssemblyIdentities)
+            {
+                string publicKeyToken = assembly.GetMetadata("PublicKeyToken");
+                if (!string.IsNullOrEmpty(publicKeyToken))
```
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: I typically try to avoid indentation when not needed.

You would write this as:
```csharp
if (string.IsNullOrEmpty(publicKeyToken))
{
    // maybe log a warning
    continue;
}

// the rest of the code.
```

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+                    XElement bindingRedirect = new XElement(ns + "bindingRedirect",
+                        new XAttribute("oldVersion", $"0.0.0.0-{assemblyVersion}"),
+                        new XAttribute("newVersion", assemblyVersion));
+                    XElement dependentAssembly = new XElement(ns + "dependentAssembly",
+                        assemblyIdentity,
+                        bindingRedirect);
+                    dependentAssemblyList.Add(dependentAssembly);
+                }
+            }
+            foreach (XElement element in dependentAssemblyList)
```
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: Why do you need this additional `foreach`? Why not directly add them to `bindingRedirectAssemblies` ?

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                bindingRedirectAssemblies.Add(element);
+            }
+            XDocument doc = new XDocument(new XElement("configuration", new XElement("runtime", bindingRedirectAssemblies)));
+            foreach (ITaskItem executable in Executables)
+            {
+                string executableName = Path.GetFileName(executable.ItemSpec);
+                using (FileStream fs = new FileStream(Path.Combine(OutputPath, executableName + ".config"), FileMode.Create))
```
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: Any reason not to use the method that takes a string as the path to the file?

```csharp
doc.Save(Path.Combine(OutputPath, executableName + ".config"))
```

I am fine with either approach.

Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: We're still dependent on a old version of XML, I can follow up on updating it.

**Comment by karajas** [comment]
PTAL

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets [hunk 15]
```diff
@@ -129,6 +129,18 @@
...
       <DestinationsForTestDir Include="@(OverridenRuntimeContent->'$(TestPath)\%(Filename)%(Extension)')" />
     </ItemGroup>
+    
+    <!-- Removing assets from TestCopyLocal that come in from P2P references in SourcesToCopyToTestDir -->
+    <ItemGroup>
+        <_FileNamesInSourcesToCopyToTestDir Include="@(SourcesToCopyToTestDir->'%(Filename)%(Extension)')">
+          <SourcePath>%(SourcesToCopyToTestDir.Identity)</SourcePath>
+        </_FileNamesInSourcesToCopyToTestDir>
+        <_FileNamesInTestCopyLocalByFileNameWithoutDuplicates Include="@(_TestCopyLocalByFileNameWithoutDuplicates->'%(Filename)%(Extension)')">
+          <SourcePath>%(_TestCopyLocalByFileNameWithoutDuplicates.SourcePath)</SourcePath>
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets]: This line does nothing.

Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets]: Moreover I don't think you need this item.  I believe _TestCopyLocalByFileNameWithoutDuplicates already has Identity of FileName+Extension and has SourcePath metadata.

Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets]: removed.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets [hunk 16]
```diff
@@ -206,6 +209,21 @@
       OutputListLocation="$(OutDir)assemblylist.txt"
       Condition="'$(UseLegacyXunitPerfRunner)'=='true'"
      />
+     
+    <!-- For Desktop Tests, we need to generate Binding Redirects. -->
+    <ItemGroup>
+      <_FilteredAssembliesForBindingRedirect Include="%(_DeduppedListOfTestAssets.SourcePath)" Condition="'%(_DeduppedListOfTestAssets.Extension)'=='.dll' or '%(IncludedFileForRunnerScript.Extension)'=='.exe'" />
```
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets]: typo will fix.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets [hunk 17]
```diff
@@ -129,6 +129,15 @@
...
       <DestinationsForTestDir Include="@(OverridenRuntimeContent->'$(TestPath)\%(Filename)%(Extension)')" />
     </ItemGroup>
+    
+    <!-- Removing assets from TestCopyLocal that come in from P2P references in SourcesToCopyToTestDir -->
+    <ItemGroup>
+        <_FileNamesInSourcesToCopyToTestDir Include="@(SourcesToCopyToTestDir->'%(Filename)%(Extension)')">
+          <SourcePath>%(SourcesToCopyToTestDir.Identity)</SourcePath>
+        </_FileNamesInSourcesToCopyToTestDir>
+        <_DeduppedListOfTestAssetsByFileName Include="@(_TestCopyLocalByFileNameWithoutDuplicates)" Exclude="@(_FileNamesInSourcesToCopyToTestDir)" />
+        <_DeduppedListOfTestAssetsByFileName Include="@(_FileNamesInSourcesToCopyToTestDir)" />
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets]: /cc @MattGal now that we have a correct merged list of files we should look at removing the hacks from the test script and the VS target that only copy over if the file doesn't already exist.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets [hunk 18]
```diff
@@ -129,6 +129,15 @@
       <SourcesToCopyToTestDir Include="@(OverridenRuntimeContent)" />
       <DestinationsForTestDir Include="@(OverridenRuntimeContent->'$(TestPath)\%(Filename)%(Extension)')" />
     </ItemGroup>
+    
+    <!-- Removing assets from TestCopyLocal that come in from P2P references in SourcesToCopyToTestDir -->
+    <ItemGroup>
+        <_FileNamesInSourcesToCopyToTestDir Include="@(SourcesToCopyToTestDir->'%(Filename)%(Extension)')">
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets]: nit: To match the convention this should also be named _SourceToCopyToTestDirByFileName

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+                XElement dependentAssembly = new XElement(ns + "dependentAssembly",
+                    assemblyIdentity,
+                    bindingRedirect);
+                bindingRedirectAssemblies.Add(dependentAssembly);
+            }
+            XDocument doc = new XDocument(new XElement("configuration", new XElement("runtime", bindingRedirectAssemblies)));
+            foreach (ITaskItem executable in Executables)
+            {
+                string executableName = Path.GetFileName(executable.ItemSpec);
+                using (FileStream fs = new FileStream(Path.Combine(OutputPath, executableName + ".config"), FileMode.Create))
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: Did you look at generating <testAssembly>.dll.config?  I believe that convention is supported by XUnit and would constrain the bindingRedirects to the isolated AppDomain that the tests run in.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets [hunk 20]
```diff
@@ -129,6 +129,15 @@
...
+          <SourcePath>%(SourcesToCopyToTestDir.Identity)</SourcePath>
+        </_FileNamesInSourcesToCopyToTestDir>
+        <_DeduppedListOfTestAssetsByFileName Include="@(_TestCopyLocalByFileNameWithoutDuplicates)" Exclude="@(_FileNamesInSourcesToCopyToTestDir)" />
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets]: nit: to me a more useful name might be "finalTestAssetsByFileName".  Something to make it clear that this is the list that folks ought to rely on for finding all the final test assemblies.  Who knows if that will actually prevent folks from further forking this in the future... but at least its an attempt.

File: src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs (hunk 19 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: I did try a version with this, I couldn't determine whether the redirect was being used from fuslog. It should be just adding the dlls to the ExecutablesForBindingRedirect itemgroup

Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/GenerateBindingRedirect.cs]: This would be applicable only for the Test.dll.config I believe? Should I go ahead and do that instead of xunit.console.exe.config?

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets (hunk 17 above)
Comment (naamunds) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets]: @ericstj See #1228.

**Comment by karajas** [comment]
This still has an issue with filtering native assemblies
